[CISupport] Autoinstall buildbot dependencies

JonWBedard opened the pull request:
#### a5d5c85284959ecd6d331a19c75b69d0e548e0de
<pre>
[CISupport] Autoinstall buildbot dependencies
<a href="https://bugs.webkit.org/show_bug.cgi?id=250604">https://bugs.webkit.org/show_bug.cgi?id=250604</a>
&lt;rdar://problem/104246259&gt;

Reviewed by Aakash Jain.

Instead of relying on machine running our unit tests having buildbot
installed, we should rely on webkitpy&apos;s autoinstaller.

* Tools/CISupport/buildbot-cmd: Added.
* Tools/CISupport/ews-build/steps.py:
(RunBuildbotCheckConfig): Invoke local &apos;buildbot-cmd&apos; script.
* Tools/CISupport/ews-build/steps_unittest.py:
* Tools/CISupport/runUnittests.py: Import buildbot from webkitpy if
autoinstall is enabled.
* Tools/Scripts/webkitpy/autoinstalled/buildbot.py: Update dependencies,
install instead of just register libraries.
* Tools/Scripts/webkitpy/autoinstalled/twisted.py: Ditto.

Canonical link: <a href="https://commits.webkit.org/262159@main">https://commits.webkit.org/262159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ec55094c6241f8c9ba7722da0538dab866edfdb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/808 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/832 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/862 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/1102 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/797 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/880 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/1102 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/816 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/880 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/862 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1042 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/834 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/880 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/862 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/1042 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/862 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/1042 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/835 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/795 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/862 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/835 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/780 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/24/builds/837 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/81 "Built successfully and passed tests") | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/24/builds/837 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/buildbot:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/python3
```
Adding a wrapper script with exact same name as `buildbot` might be extremely confusing. Better to avoid same name. 

Further, do we need this wrapper script, and change 'buildbot checkconfig' invocations? It might be better to add a build-step to install the required packages (or build-step to invoke a script which does autoinstall) and run any buildbot commands normally after that. That way if anyone try to reproduce any issue manually on the bot, that would also work.

JonWBedard replied:
This script actually _is_ the `buildbot` script, just with autoinstall hooks.

This is an explicit attempt to avoid installing the required packages because installing the required packages is not intuitive. Getting autoinstall to work with our unit tests means that we're very close to having it work with `buildbot` commands too, we just have to pick a name for the entry script, basically.

aj062 replied:
ok. we should rename it to something explicit for now and use it only in CI. If it works smoothy for a while, then we can consider if we should rename it to buildbot or not.

aj062 commented on Tools/CISupport/runUnittests.py:
```diff
@@ -31,4 +27,15 @@ def run_unittests(test_discovery_path):
 
     path = os.path.abspath(relative_path)
     assert os.path.isdir(path), '{} is not a directory. Please specify a valid directory'.format(path)
+
```
Not sure if we want to autoinstall even in the case when people are running this script (unit-tests) locally on their computers. 

Also would it autoinstall buildbot pip package with `--no-binary :all` flags? Usually we need that for proper buildbot installation.

JonWBedard replied:
I think we _definately_ want to autoinstall when folks are running locally. One of the biggest problems I have when working on buildbot (switching machines, in particular) is that we don't have a reliable way to install buildbot and its dependencies.

`--no-binary :all:` is basically opting out of `wheel` installs, which `autoinstall` does by default. So yes, we're doing `--no-binary :all:` (except for the packages which really need to be binary, like `cryptography` and friends, but also, we don't end up installing those packages with `--no-binary :all:` in practice because they're usually already installed when installing buildbot dependencies...that being said, the "usually" here has frequently tripped me up when setting up a new machine)

aj062 replied:
My main concern is reliability, whether autoinstall would be reliably able to install it. 
Otherwise we are going to create a mess, when autoinstall decides to ignore the package already installed on the machine, tries to install buildbot and fail. Effectively causing users problems even though they had valid buildbot version already installed.

aj062 commented on Tools/CISupport/runUnittests.py:
```diff
@@ -21,14 +18,44 @@ def run_unittests(test_discovery_path):
+        help='Relative path of directory to run Python unit tests in',
+    )
+    parser.add_argument(
+        '--no-autoinstall',
```
Let's keep the default behavior to not autoinstall for now (with the parameter --autoinstall passed for auto installation). If it is working fine without any issues in production, we can then enable autoinstall by default.

aj062 commented on Tools/CISupport/buildbot-cmd:
```diff
@@ -0,0 +1,54 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
Would be good to add a reference to the original buildbot file url somewhere in this file.

aj062 reviewed (approved):
r+ with comments.

JonWBedard committed:
[CISupport] Autoinstall buildbot dependencies
https://bugs.webkit.org/show_bug.cgi?id=250604
<rdar://problem/104246259>

Reviewed by Aakash Jain.

Instead of relying on machine running our unit tests having buildbot
installed, we should rely on webkitpy's autoinstaller.

* Tools/CISupport/buildbot-cmd: Added.
* Tools/CISupport/ews-build/steps.py:
(RunBuildbotCheckConfig): Invoke local 'buildbot-cmd' script.
* Tools/CISupport/ews-build/steps_unittest.py:
* Tools/CISupport/runUnittests.py: Import buildbot from webkitpy if
autoinstall is enabled.
* Tools/Scripts/webkitpy/autoinstalled/buildbot.py: Update dependencies,
install instead of just register libraries.
* Tools/Scripts/webkitpy/autoinstalled/twisted.py: Ditto.

Canonical link: https://commits.webkit.org/262159@main
